### PR TITLE
deps: updates for bazel 8 compatibility

### PR DESCRIPTION
**Pull request by rickeylev**
Various changes to support Bazel 8. An important note is dependencies have forced
us to change the versions of Bazel we support.

Summary of changes:
* rules_cc 0.0.14: Releases after 0.0.9 have some Bazel 8 fixes, but also broke
  some things. Things seemed to have settled by 0.0.14.
* protobuf 29.0-rc1: Technically 28.0 works, however:
  1. 29.0-rc1 is coming via a transitive dependency anyways, and
  2. In protobuf 28.0, compile warnings are treated as errors, which our Debian CI
     respects (and thus fails), while other platforms ignore.
* stardoc 0.7.1: Fixes an issue with Bazel 8 and stardoc using empty globs.
* Bazel 7.4 is now the minimum supported Bazel version. This requirements comes via
  dependencies.
* Drop Bazel 6 bzlmod support. This requirement comes via dependencies.
* Add a presubmit job for `last_rc` Bazel (currently the 8.x RC).
* Use a local patch so Gazelle works with Bazel 8. This can be removed
  once https://github.com/bazel-contrib/bazel-gazelle/issues/1959 is fixed and released.
* Fix a `$(rpathlocation)` call in bootstrap tests.
* Update bzl_library deps after upgrading deps: the set of targets that provide
  bzl sources changed in rules_cc and protobuf in these newer versions.

Sorting this all out and finding the right combination of dependency versions was
fairly involved. The details of that are in
https://github.com/bazelbuild/rules_python/issues/2378.

Work towards https://github.com/bazelbuild/rules_python/issues/2378, https://github.com/bazelbuild/rules_python/issues/2387

### CODE REVIEW

**Comment by aignas** [comment]
Ah, I read the PR I linked - I somewhat think that maybe we could couple the PRs together and see if we can remove the extra protobuf deps that we had to add. `rules_rust` in our examples looks extremely concerning.

**Comment by aignas** [comment]
nit: I think we should also resolve the following warnings in the `rules_python` repo:
```
WARNING: For repository 'bazel_features', the root module requires module version bazel_features@1.9.1, but got bazel_features@1.19.0 in the resolved dependency graph. Please update the version in your MODULE.bazel or set --check_direct_dependencies=off
WARNING: For repository 'platforms', the root module requires module version platforms@0.0.4, but got platforms@0.0.10 in the resolved dependency graph. Please update the version in your MODULE.bazel or set --check_direct_dependencies=off
WARNING: For repository 'rules_proto', the root module requires module version rules_proto@6.0.0-rc1, but got rules_proto@6.0.2 in the resolved dependency graph. Please update the version in your MODULE.bazel or set --check_direct_dependencies=off
WARNING: For repository 'com_google_protobuf', the root module requires module version protobuf@29.0-rc1, but got protobuf@29.0-rc2 in the resolved dependency graph. Please update the version in your MODULE.bazel or set --check_direct_dependencies=off
```

File: MODULE.bazel [hunk 1]
```diff
@@ -83,8 +83,18 @@ bazel_dep(name = "bazel_ci_rules", version = "1.0.0", dev_dependency = True)
 # Extra gazelle plugin deps so that WORKSPACE.bzlmod can continue including it for e2e tests.
 # We use `WORKSPACE.bzlmod` because it is impossible to have dev-only local overrides.
 bazel_dep(name = "rules_go", version = "0.41.0", dev_dependency = True, repo_name = "io_bazel_rules_go")
-bazel_dep(name = "gazelle", version = "0.33.0", dev_dependency = True, repo_name = "bazel_gazelle")
 bazel_dep(name = "rules_python_gazelle_plugin", version = "0", dev_dependency = True)
+bazel_dep(name = "gazelle", version = "0.33.0", dev_dependency = True, repo_name = "bazel_gazelle")
+single_version_override(
```
[the remark this answers is not in the document]
Comment (rickeylev) [reply MODULE.bazel:88]: Correct. We'll be able to drop this patch soon. Gazelle has fixed the bug at head and is just working through some BCR release issues before they can release.

File: examples/multi_python_versions/MODULE.bazel [hunk 2]
```diff
@@ -59,3 +59,6 @@ pip.parse(
 
 # example test dependencies
 bazel_dep(name = "rules_shell", version = "0.2.0", dev_dependency = True)
+
+# Only needed to make rules_python's CI happy
+bazel_dep(name = "rules_java", version = "8.3.0")
```
[the remark this answers is not in the document]
Comment (rickeylev) [reply examples/multi_python_versions/MODULE.bazel:65]: I updated the comment to explain why it's here.

The example sets `--java_runtime_version=remotejdk_11`, which is broken with Bazel 8 until rules_java 8.3.

File: python/private/BUILD.bazel [hunk 3]
```diff
@@ -632,7 +632,16 @@ bzl_library(
 # @rules_cc does not offer a bzl_library target for @rules_cc//cc:defs.bzl
 bzl_library(
     name = "rules_cc_srcs_bzl",
-    srcs = ["@rules_cc//cc:bzl_srcs"],
+    srcs = [
+        # It appears something in the CC rules loads cc_proto_library, but
+        # where isn't clear.
+        "@com_google_protobuf//:bzl_srcs",
```
[the remark this answers is not in the document]
Comment (rickeylev) [reply python/private/BUILD.bazel:640]: Hm, yeah, removing defs.bzl might fix this (depends on if the more specific targets properly specify their bzl_libraries :sweat_smile: )

I wrote this comment before (1) I unwound the whole defs.bzl removing cc_proto_library thing, and (2) I gave up trying to make bazel 6 with bzlmod work.

I've updated the comment to better reflect this.

File: examples/bzlmod/MODULE.bazel [hunk 4]
```diff
@@ -12,10 +12,14 @@ local_path_override(
...
-bazel_dep(name = "rules_proto", version = "5.3.0-21.7")
+bazel_dep(name = "rules_proto", version = "6.0.0-rc1")
 
 # (py_proto_library specific) Add the protobuf library for well-known types (e.g. `Any`, `Timestamp`, etc)
-bazel_dep(name = "protobuf", version = "24.4", repo_name = "com_google_protobuf")
+bazel_dep(name = "protobuf", version = "27.0", repo_name = "com_google_protobuf")
+bazel_dep(name = "rules_java", version = "8.3.0")
+
+##bazel_dep(name = "rules_rust", version = "0.45.1")
+bazel_dep(name = "rules_rust", version = "0.54.1")
```
[the remark this answers is not in the document]
Comment (rickeylev) [reply examples/bzlmod/MODULE.bazel:28]: This is to fix the module.lock file with host os/arch specific content problem I mentioned in Slack. Earlier versions of rules_rust don't set the os/arch specific module extension settings, so a hash relating to "rust host tools" differs between e.g. linux and mac. Later versions set the flag, so while each platform will generate a different lockfile from scratch, they're OK with having the extra os-specific content in their lockfiles (so we just have to copy/paste the diff CI reports into the lockfile going forward).

I updated the comment to reflect this.

**Comment by rickeylev** [comment]
> The only real question is the defs.bzl usage from rules_cc. I remember there was a PR here #2293, but I can still find usage of defs.bzl in our git repo in this PR.

Yeah, this whole part is a mess. Our usage of rules_cc/defs.bzl is OK, sort of -- deprecated, but not broken, per-se. The problem comes when other dependencies come into play. Basically, rules_cc removed a symbol (cc_proto_library) from defs.bzl. However, some things still loaded that symbol, so fixes were sent out. While all this is going on, various versions of modules were released with various problems with Bazel 6, Bazel 8, and/or the removal of that symbol.

So, our usage of rules_cc/defs.bzl is only a problem depending on what combinations of what dependencies you end up with. Us switching off rules_cc/defs.bzl makes it less likely to occur (since we wouldn't be a potential cause of loading that file when an incompatible dependency is being used).

re: coupling this PR and the defs.bzl removal: Well, I was going to say this PR has been a chore and I don't want to expand it's scope, but now CI is failing with yet-another-rules-cc-defs.bzl looking failure. So maybe it'll be easier to put those in one PR. I'll have a look at that failure to see which is gonna be easier.

re: module versions:
* bazel_features and platforms are coming from dev dependencies, I think. I don't see those warnings when building the bzlmod example.
* rules_proto, protobuf: I've updated their versions

**Comment by rickeylev** [comment]
> CI is failing with yet-another-rules-cc-defs.bzl looking failure

The failure was: Bazel 9 with workspace was failing because Bazel 9 removes `native.cc_proto_library`

I've "fixed" this by removing the bazel-in-bazel integration tests for rolling, A bzlmod invocation is OK because it uses protobuf 29.0-rc1+, which has a fix for that. However, Bazel 9 with workspace is using an earlier protobuf version. I tried updating that, but it looks like it has further dependencies that need to be setup and/or upgraded.

Given that Bazel 9 is a ways off, I think it makes sense to temporarily disable that test. I'll need some time to try and figure out the necessary WORKSPACE updates that need to be made.

File: examples/multi_python_versions/MODULE.bazel (hunk 2 above)
[the remark this answers is not in the document]
Comment (aignas) [reply examples/multi_python_versions/MODULE.bazel:65]: And if I remember correctly this is only to make the coverage collection not depend on system Java. So technically all of our consumers will need to add this dep if they want coverage.

**Comment by aignas** [comment]
Thanks for all of this untangling of the mess. This is sure a bumpy one - I remember bazel 7 not being nearly as bad.

**Review by aignas — APPROVED** [review]
Approving since the CI is green and we can go bit by bit fixing all of the remaining things. My main concerns are:
* Can we remove some of the deps added here by switching away from `rules_cc:defs.bzl`.
* The `java` flag in the `bazelrc` was there for `bazel coverage` support and I don't understand if we can safely remove it and the coverage will just work or we have to document it to the example code readers, that they need to have that dep as well?

Regardless of the outcome of the above, this is moving in the right direction.

**Pull request merged**